src/plugins/api/index_pattern_endpoints.js

```diff
@@ -0,0 +1,221 @@
+      error instanceof esErrors.ServiceUnavailable ||
+      error instanceof esErrors.NoConnections ||
+      error instanceof esErrors.RequestTimeout) {
+      return Boom.serverTimeout();
+    } else if (error instanceof esErrors.Conflict) {
+      return Boom.conflict();
+    } else if (error instanceof esErrors[403]) {
+      return Boom.forbidden();
+    } else if (error instanceof esErrors.NotFound) {
+      return Boom.notFound();
+    } else if (error instanceof esErrors.BadRequest || error instanceof TypeError) {
+      return Boom.badRequest(error);
+    } else {
+      return error;
+    }
+  };
+
+  server.route({
+    path: '/api/index-patterns',
+    method: 'GET',
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
+
+      client.search({
+        index: '.kibana',
+        type: 'index-pattern',
+        body: {
+          query: {
+            match_all: {}
+          }
+        }
+      }).then(function parseResults(results) {
+        const hits = results.hits.hits;
+        return _.map(hits, (patternHit) => {
+          patternHit._source.fields = JSON.parse(patternHit._source.fields);
+          return patternHit._source;
+        });
+      }).then((patterns) => {
+        return Promise.map(patterns, (pattern) => {
+          return getMappings(pattern.title, client).catch(() => {
+            return {};
+          }).then((mappings) => {
+            return stitchPatternAndMappings(pattern, mappings);
+          });
+        });
+      }).then(removeDeprecatedFieldProps)
+        .then((patterns) => {
+          reply(patterns);
+        }, function (error) {
+          reply(handleESError(error));
+        });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns/{id}',
+    method: 'GET',
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
+      let pattern = req.params.id;
+
+      Promise.join(
+        client.get({
+          index: '.kibana',
+          type: 'index-pattern',
+          id: req.params.id
+        }).then((result) => {
+          result._source.fields = JSON.parse(result._source.fields);
+          return result._source;
+        }),
+        getMappings(pattern, client),
+        stitchPatternAndMappings
+      ).then(removeDeprecatedFieldProps)
+        .then(function (pattern) {
+          reply(pattern);
+        }, function (error) {
+          reply(handleESError(error));
+        });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns',
+    method: 'POST',
+    handler: function (req, reply) {
+      const client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
+      const mappings = _.omit(_.mapValues(_.indexBy(req.payload.fields, 'name'), (value) => {
+        return value.mapping;
+      }), _.isUndefined);
```

Comment:
For readability I usually prefer something like this when I use a lot of lodash methods:

```javascript
_(req.payload.fields)
    .indexBy('name')
    .mapValues(value => value.mapping)
    .omit(_.isUndefined)
    .value();
```